### ISSUE TITLE
Issue with CollectionInputFilter and nested collections

### ISSUE THREAD

**Issue by adamlundrigan** [zendframework/zendframework#6304]
TL;DR: CollectionInputFilter reuses the same InputFilter instance to validate each element in the collection, leading to a surprise when those elements contain collections themselves. 

-------

I have a form setup of this structure:

```
Form <-- Base Fieldset <-- Collection <-- Fieldset <-- Collection
```

Each element of the Collection on the base fieldset is a fieldset which itself contains a collection.  This works fine on the form side, and when I attach the DoctrineModule hydrator I get back a fully-populated entity graph.  However, when I started building the matching InputFilter hierarchy using InputFilter and CollectionInputFilter classes:

```
InputFilter <-- InputFilter <-- CollectionInputFilter <-- InputFilter <-- CollectionInputFilter
```

the validation always fails.  Digging into CollectionInputFilter I found that [isValid re-uses the same `$inputFilter` instance](https://github.com/zendframework/zf2/blob/master/library/Zend/InputFilter/CollectionInputFilter.php#L174) to validate each member of the collection.  This doesn't work as the inner CollectionInputFilter instance bleeds data between the elements of the outer CollectionInputFilter.  

An example:  Say I have this data structure

```
[
    'form' => [
        'fieldset' => [
            'outerCollection' => [
                0 => [
                    'innerCollection' => [
                        0 => 'foo',
                        1 => 'bar',
                        2 => 'baz',
                    ],
                ],
                1 => [
                    'innerCollection' => [
                        0 => 'baaat',
                    ],
                ],
            ],            
        ],
    ],
],
```

If I override the `isValid` method of outerCollection and `var_dump($this->getValues())`, I will get something that looks like this:

```
[
    0 => [
        'innerCollection' => [
            0 => 'foo',
            1 => 'bar',
            2 => 'baz',
        ],
    ],
    1 => [
        'innerCollection' => [
            0 => 'baaat',
            1 => 'bar',
            2 => 'baz',
        ],
    ],
],
```

In my particular use case I want to ensure that the contents of the inner collections are globally unique (a value can reside in only one innerCollection at a time), and so the validation will fail because `bar` and `baz` are in both `innerCollection`s

For my use case this 'hack' fixes the issue, since it disables the InputFilter reuse:
Remove [Line 150](https://github.com/zendframework/zf2/blob/3f870db2455e6c3dac746865dfca37333ec8c917/library/Zend/InputFilter/CollectionInputFilter.php#L150) and add the following above [Line 174](https://github.com/zendframework/zf2/blob/3f870db2455e6c3dac746865dfca37333ec8c917/library/Zend/InputFilter/CollectionInputFilter.php#L174):

```
$inputFilter = clone $this->getInputFilter();
```

**Comment by GeeH** [comment]
This issue has been closed as part of the bug migration program as outlined here - http://framework.zend.com/blog/2016-04-11-issue-closures.html

**Comment by adamlundrigan** [comment]
This was fixed in https://github.com/zendframework/zendframework/pull/6412
